javascript/ql/lib/semmle/javascript/frameworks/ApolloGraphQL.qll

```diff
@@ -0,0 +1,56 @@
+/**
+ * Provides classes for working with Apollo GraphQL connectors.
+ */
+
+import javascript
+
+/** Provides classes modeling concepts of Apollo GraphQL. */
+module ApolloGraphQL {
+  /** A string-valued expression that is interpreted as a Apollo GraphQL query. */
+  abstract class GraphQLString extends DataFlow::Node { }
+
+  /** A string-valued expression that is interpreted as a Apollo GraphQL query. */
+  abstract class ApolloGraphQLServer extends DataFlow::Node { }
+}
+
+/**
+ * Provides classes modeling the apollo packages [@apollo/server](https://npmjs.com/package/@apollo/server`)
+ */
+private module Apollo {
+  /** Get an instanceof of `Apollo` */
+  private API::Node apollo() {
+    result =
+      API::moduleImport([
+          "@apollo/server", "apollo/server", "@apollo/apollo-server-express",
+          "@apollo/apollo-server-core", "apollo-server", "apollo-server-express"
+        ]).getMember("ApolloServer")
+  }
+
+  /** Get an instanceof of `gql` */
+  private API::Node gql() {
+    result =
+      API::moduleImport([
+          "@apollo/server", "apollo/server", "@apollo/apollo-server-express",
+          "@apollo/apollo-server-core", "apollo-server", "apollo-server-express"
+        ]).getMember("gql")
+  }
+
+  /** A string that is interpreted as a GraphQL query by a `octokit` package. */
+  private class ApolloGraphQLString extends GraphQL::GraphQLString {
+    ApolloGraphQLString() { this = gql().getACall() }
+  }
```

Comment:
Hi Erik 😄 ! That part is not finished because I wanted to ask how it could be done because I couldn't get it to work. I don't know why using `.getACall().getArgument(0)` works with `execFile` and not with `gql`. Any ideas?

```js
const { execFile } = require("child_process");
const { ApolloServer, gql } = require("apollo-server");

execFile("whoami"); 

const typeDefs = gql`
  type Query {
    hello: String
  }
`;
```

Reply:
Yes, tagged template strings are not currently seen as calls in our analysis, and you're not the only one that's noticed this limitation recently.  
The proper fix is to get tagged template strings added to `DataFlow::CallNode` (which involves adding a new case in `DataFlow.qll` that extends `CallNodeDef`).  

Your use with API-graphs should just magically work after tagged template strings are added as `DataFlow::CallNode`s.   

I'll look into fixing that limitation of our analysis later, but it might take a while before I get started on that. 

If you're feeling adventurous you can try it yourself (that would definitely belong in a separate PR).

Reply:
I just opened a PR to recognize tagged template literals as a `DataFlow::CallNode`: https://github.com/github/codeql/pull/14405

 I suggest you take a look at that PR and incorporate it into your PR after my PR has been merged.
 There is no rush, so just wait until my PR has been merged.

Reply:
Sure! Thanks :) 👍

Reply:
https://github.com/github/codeql/pull/14405 has been merged, so you can incorporate that into this PR.